File: mtr_app/services/mtr_service.py
"""
MTR Service Layer.
Connects to enasr-geospatial.db SQLite database, enriches MTR data,
computes corridor geometry, and provides format conversions (XML, Fixed, GeoJSON, Plain-English).
"""

import sqlite3
import json
import os
from typing import List, Dict, Any, Optional
from mtr_app.generators.corridor_calc import (
    calculate_bearing,
    calculate_distance_nm,
    to_arinc_dms,
    to_human_dms,
    generate_corridor_polygon
)
from mtr_app.generators.arinc424_xml import (
    generate_arinc424_xml,
    generate_single_route_xml,
    validate_arinc424_xml,
    decode_to_plain_english
)
from mtr_app.generators.arinc424_fixed import (
    generate_mtr_fixed_records,
    generate_all_fixed_records
)

from mtr_app.services.supabase_service import SupabaseService, load_env_file

DEFAULT_DB_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "enasr-geospatial.db")


class MTRService:
    def __init__(self, db_path: Optional[str] = None):
        self.db_path = db_path or DEFAULT_DB_PATH
        self._ensure_db_initialized()
        self.supabase = SupabaseService()

    def _get_connection(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path)
        conn.execute("PRAGMA foreign_keys = ON;")
        conn.row_factory = sqlite3.Row
        return conn

    def _ensure_db_initialized(self):
        """Verify MTR tables exist in the database."""
        conn = self._get_connection()
        try:
            cursor = conn.cursor()
            cursor.execute("SELECT name FROM sqlite_master WHERE type='table' AND name='mtr_routes'")
            if not cursor.fetchone():
                # Read schema from enasr-schema.sql if needed
                schema_path = os.path.join(os.path.dirname(self.db_path), "enasr-schema.sql")
                if os.path.exists(schema_path):
                    with open(schema_path, "r", encoding="utf-8") as f:
                        conn.executescript(f.read())
        finally:
            conn.close()

    def list_routes(self, route_type: Optional[str] = None) -> List[Dict[str, Any]]:
        """List all MTR routes with summary statistics (Supabase if configured, SQLite fallback)."""
        if self.supabase.is_configured():
            try:
                sb_routes = self.supabase.list_routes(route_type=route_type)
                if sb_routes:
                    return sb_routes
            except Exception as e:
                print(f"[Warning] Supabase fetch failed: {e}. Falling back to SQLite.")

        conn = self._get_connection()
        try:
            cursor = conn.cursor()
            if route_type:
                cursor.execute("""
                    SELECT r.*, COUNT(s.segment_id) as waypoint_count 
                    FROM mtr_routes r 
                    LEFT JOIN mtr_segments s ON r.route_id = s.route_id
                    WHERE r.route_type = ?
                    GROUP BY r.route_id
                    ORDER BY r.route_id
                """, (route_type.upper(),))
            else:
                cursor.execute("""
                    SELECT r.*, COUNT(s.segment_id) as waypoint_count 
                    FROM mtr_routes r 
                    LEFT JOIN mtr_segments s ON r.route_id = s.route_id
                    GROUP BY r.route_id
                    ORDER BY r.route_id
                """)
            routes = [dict(row) for row in cursor.fetchall()]
            return routes
        finally:
            conn.close()

    def get_route_details(self, route_id: str) -> Optional[Dict[str, Any]]:
        """Fetch full details for an MTR route, enriched with segment bearings and corridor polygon."""
        route = None
        raw_segments = []

        if self.supabase.is_configured():
            try:
                sb_route = self.supabase.get_route_details(route_id)
                if sb_route:
                    route = sb_route
                    raw_segments = route.get("segments", [])
            except Exception as e:
                print(f"[Warning] Supabase route fetch failed: {e}. Falling back to SQLite.")

        if not route:
            conn = self._get_connection()
            try:
                cursor = conn.cursor()
                cursor.execute("SELECT * FROM mtr_routes WHERE route_id = ?", (route_id,))
                row = cursor.fetchone()
                if not row:
                    return None
                
                route = dict(row)
                cursor.execute("SELECT * FROM mtr_segments WHERE route_id = ? ORDER BY sequence_num", (route_id,))
                raw_segments = [dict(s) for s in cursor.fetchall()]
            finally:
                conn.close()

        # Enrich segments
        enriched_segments = []
        cumulative_nm = 0.0

        for i, seg in enumerate(raw_segments):
            seg_dict = dict(seg)
            lat1 = seg_dict.get("latitude_dec", 0.0)
            lon1 = seg_dict.get("longitude_dec", 0.0)

            # Determine next coordinates if missing
            lat2 = seg_dict.get("next_lat_dec")
            lon2 = seg_dict.get("next_lon_dec")
            if (lat2 is None or lon2 is None) and i + 1 < len(raw_segments):
                lat2 = raw_segments[i + 1].get("latitude_dec")
                lon2 = raw_segments[i + 1].get("longitude_dec")
                seg_dict["next_lat_dec"] = lat2
                seg_dict["next_lon_dec"] = lon2
                if not seg_dict.get("next_point_name"):
                    seg_dict["next_point_name"] = raw_segments[i + 1].get("point_name")

            # Distance and Bearing
            dist = seg_dict.get("segment_distance_nm")
            if lat2 is not None and lon2 is not None:
                bearing = calculate_bearing(lat1, lon1, lat2, lon2)
                seg_dict["bearing_deg"] = bearing
                if dist is None:
                    dist = calculate_distance_nm(lat1, lon1, lat2, lon2)
                    seg_dict["segment_distance_nm"] = dist

            dist_val = dist if dist is not None else 0.0
            seg_dict["cumulative_distance_nm"] = round(cumulative_nm, 1)
            cumulative_nm += dist_val

            # Coordinate strings
            arinc_lat, arinc_lon = to_arinc_dms(lat1, lon1)
            human_lat, human_lon = to_human_dms(lat1, lon1)
            seg_dict["arinc_lat"] = arinc_lat
            seg_dict["arinc_lon"] = arinc_lon
            seg_dict["human_lat"] = human_lat
            seg_dict["human_lon"] = human_lon

            enriched_segments.append(seg_dict)

        route["segments"] = enriched_segments
        route["total_distance_nm"] = round(cumulative_nm, 1)

        # Generate corridor polygon
        corridor_poly = generate_corridor_polygon(enriched_segments)
        route["corridor_polygon"] = corridor_poly

        return route

    def get_route_geojson(self, route_id: str) -> Optional[Dict[str, Any]]:
        """Returns a GeoJSON FeatureCollection for a single MTR route with Centerline, Corridor, and Waypoints."""
        route = self.get_route_details(route_id)
        if not route:
            return None

        features = []
        segments = route.get("segments", [])

        # 1. Centerline LineString
        coords = []
        for s in segments:
            coords.append([s["longitude_dec"], s["latitude_dec"]])
        if segments and segments[-1].get("next_lon_dec") is not None:
            coords.append([segments[-1]["next_lon_dec"], segments[-1]["next_lat_dec"]])

        features.append({
            "type": "Feature",
            "id": f"{route_id}-centerline",
            "properties": {
                "feature_type": "CENTERLINE",
                "route_id": route_id,
                "route_type": route.get("route_type"),
                "route_name": route.get("route_name"),
                "agency": route.get("originating_agency"),
                "floor_alt_ft": route.get("floor_alt_ft"),
                "ceiling_alt_ft": route.get("ceiling_alt_ft"),
                "width_nm": route.get("route_width_nm"),
                "total_distance_nm": route.get("total_distance_nm")
            },
            "geometry": {
                "type": "LineString",
                "coordinates": coords
            }
        })

        # 2. Corridor Ribbon Polygon
        corridor_coords = route.get("corridor_polygon", [])
        if corridor_coords:
            features.append({
                "type": "Feature",
                "id": f"{route_id}-corridor",
                "properties": {
                    "feature_type": "CORRIDOR_RIBBON",
                    "route_id": route_id,
                    "width_nm": route.get("route_width_nm"),
                    "floor_alt_ft": route.get("floor_alt_ft"),
                    "ceiling_alt_ft": route.get("ceiling_alt_ft")
                },
                "geometry": {
                    "type": "Polygon",
                    "coordinates": [corridor_coords]
                }
            })

        # 3. Waypoint Points
        for s in segments:
            features.append({
                "type": "Feature",
                "id": f"{route_id}-{s['point_name']}",
                "properties": {
                    "feature_type": "WAYPOINT",
                    "route_id": route_id,
                    "point_name": s["point_name"],
                    "point_type": s["point_type"],
                    "sequence_num": s["sequence_num"],
                    "min_alt_ft": s.get("min_alt_ft"),
                    "max_alt_ft": s.get("max_alt_ft"),
                    "human_coords": f"{s['human_lat']}, {s['human_lon']}",
                    "arinc_coords": f"{s['arinc_lat']}, {s['arinc_lon']}"
                },
                "geometry": {
                    "type": "Point",
                    "coordinates": [s["longitude_dec"], s["latitude_dec"]]
                }
            })

        return {
            "type": "FeatureCollection",
            "name": f"MTR_{route_id}",
            "route_id": route_id,
            "features": features
        }

    def get_all_routes_geojson(self) -> Dict[str, Any]:
        """Returns a unified GeoJSON FeatureCollection of all MTR routes."""
        all_routes = self.list_routes()
        unified_features = []
        for r_summary in all_routes:
            r_geo = self.get_route_geojson(r_summary["route_id"])
            if r_geo:
                unified_features.extend(r_geo.get("features", []))

        return {
            "type": "FeatureCollection",
            "name": "eNASR_All_Military_Training_Routes",
            "features": unified_features
        }

    def get_arinc424_xml(self, route_id: Optional[str] = None, pretty: bool = True) -> str:
        """Returns ARINC 424-23 compliant XML for a single route or all routes."""
        if route_id:
            route = self.get_route_details(route_id)
            if not route:
                return f"<!-- Route '{route_id}' not found -->"
            return generate_single_route_xml(route, pretty=pretty)
        else:
            routes = [self.get_route_details(r["route_id"]) for r in self.list_routes()]
            routes = [r for r in routes if r is not None]
            return generate_arinc424_xml(routes, pretty=pretty)

    def get_arinc424_fixed(self, route_id: Optional[str] = None) -> str:
        """Returns ARINC 424 132-character fixed records."""
        if route_id:
            route = self.get_route_details(route_id)
            if not route:
                return f"// Route '{route_id}' not found\n"
            records = generate_mtr_fixed_records(route)
            return "\n".join(records)
        else:
            routes = [self.get_route_details(r["route_id"]) for r in self.list_routes()]
            routes = [r for r in routes if r is not None]
            return generate_all_fixed_records(routes)

    def get_plain_english(self, route_id: str) -> str:
        """Returns human-readable plain-English operational flight brief."""
        route = self.get_route_details(route_id)
        if not route:
            return f"Route '{route_id}' not found."
        return decode_to_plain_english(route)

    def create_or_update_route(self, route_data: Dict[str, Any]) -> Dict[str, Any]:
        """
        Creates or updates an MTR route and its segments in the SQLite database.
        """
        route_id = route_data.get("route_id", "").strip().upper()
        if not route_id:
            raise ValueError("route_id is required")

        route_type = route_data.get("route_type", "IR").strip().upper()
        route_name = route_data.get("route_name", f"{route_id} Training Route")
        originating_agency = route_data.get("originating_agency", "USAF / Command")
        artcc_facility = route_data.get("artcc_facility", "ARTCC")
        operating_hours = route_data.get("operating_hours", "CONTINUOUS")
        floor_alt_ft = int(route_data.get("floor_alt_ft", 100))
        ceiling_alt_ft = int(route_data.get("ceiling_alt_ft", 15000))
        route_width_nm = float(route_data.get("route_width_nm", 10.0))
        status = route_data.get("status", "ACTIVE")
        segments = route_data.get("segments", [])

        # Build WKT LineString
        wkt_coords = []
        for s in segments:
            wkt_coords.append(f"{s['longitude_dec']} {s['latitude_dec']}")
        wkt = f"LINESTRING({', '.join(wkt_coords)})" if wkt_coords else None

        if self.supabase.is_configured():
            try:
                self.supabase.create_or_update_route(route_data)
            except Exception as e:
                print(f"[Warning] Supabase sync failed: {e}")

        conn = self._get_connection()
        try:
            cursor = conn.cursor()
            cursor.execute("""
                INSERT INTO mtr_routes (
                    route_id, route_type, route_name, originating_agency, artcc_facility,
                    floor_alt_ft, ceiling_alt_ft, route_width_nm, status, geometry_wkt
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                ON CONFLICT(route_id) DO UPDATE SET
                    route_type = excluded.route_type,
                    route_name = excluded.route_name,
                    originating_agency = excluded.originating_agency,
                    artcc_facility = excluded.artcc_facility,
                    floor_alt_ft = excluded.floor_alt_ft,
                    ceiling_alt_ft = excluded.ceiling_alt_ft,
                    route_width_nm = excluded.route_width_nm,
                    status = excluded.status,
                    geometry_wkt = excluded.geometry_wkt
            """, (
                route_id, route_type, route_name, originating_agency, artcc_facility,
                floor_alt_ft, ceiling_alt_ft, route_width_nm, status, wkt
            ))

            # Delete old segments and insert new ones
            cursor.execute("DELETE FROM mtr_segments WHERE route_id = ?", (route_id,))
            for i, s in enumerate(segments):
                seq_num = s.get("sequence_num", i + 1)
                pt_name = s.get("point_name", f"PT_{seq_num}")
                pt_type = s.get("point_type", "ENTRY" if seq_num == 1 else "WAYPOINT")
                lat1 = float(s.get("latitude_dec", 0.0))
                lon1 = float(s.get("longitude_dec", 0.0))
                min_alt = int(s.get("min_alt_ft", floor_alt_ft))
                max_alt = int(s.get("max_alt_ft", ceiling_alt_ft))
                w_left = float(s.get("width_left_nm", route_width_nm / 2.0))
                w_right = float(s.get("width_right_nm", route_width_nm / 2.0))
                next_pt = s.get("next_point_name")
                next_lat = s.get("next_lat_dec")
                next_lon = s.get("next_lon_dec")
                seg_dist = s.get("segment_distance_nm")

                seg_wkt = None
                if next_lat is not None and next_lon is not None:
                    seg_wkt = f"LINESTRING({lon1} {lat1}, {next_lon} {next_lat})"

                cursor.execute("""
                    INSERT INTO mtr_segments (
                        route_id, sequence_num, point_name, point_type,
                        latitude_dec, longitude_dec, min_alt_ft, max_alt_ft,
                        width_left_nm, width_right_nm, next_point_name,
                        segment_distance_nm, geometry_wkt
                    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """, (
                    route_id, seq_num, pt_name, pt_type,
                    lat1, lon1, min_alt, max_alt,
                    w_left, w_right, next_pt,
                    seg_dist, seg_wkt
                ))

            conn.commit()
            return self.get_route_details(route_id) or {}
        finally:
            conn.close()

File: src/services/mtrService.ts
import fs from "fs";
import path from "path";
import { MTRRoute, MTRRouteSummary, MTRSegment } from "../types.js";
import {
  calculateBearing,
  calculateDistanceNm,
  destinationPoint,
  toArincDms,
  toHumanDms,
  generateCorridorPolygon,
} from "../generators/corridorCalc.js";
import {
  generateArinc424Xml,
  generateSingleRouteXml,
  decodeToPlainEnglish,
} from "../generators/arinc424Xml.js";
import {
  generateMtrFixedRecords,
  generateAllFixedRecords,
} from "../generators/arinc424Fixed.js";

export class MTRService {
  private routes: Map<string, MTRRoute> = new Map();

  constructor() {
    this.loadInitialData();
  }

  private loadInitialData() {
    try {
      const candidates = [
        path.join(process.cwd(), "initial_routes.json"),
        path.join(process.cwd(), "enasr-mtr-routes.json"),
      ];
      let loaded = false;

      for (const p of candidates) {
        if (fs.existsSync(p)) {
          const raw = fs.readFileSync(p, "utf-8");
          const parsed = JSON.parse(raw);
          if (Array.isArray(parsed)) {
            for (const r of parsed) {
              this.enrichAndSaveRoute(r);
            }
            loaded = true;
            console.log(`Loaded ${this.routes.size} routes from ${path.basename(p)}`);
            break;
          } else if (parsed.features && Array.isArray(parsed.features)) {
            // GeoJSON feature collection
            for (const f of parsed.features) {
              const props = f.properties || {};
              const coords = f.geometry?.coordinates || [];
              const segments: MTRSegment[] = coords.map((c: number[], idx: number) => ({
                sequence_num: idx + 1,
                point_name: `PT_${idx + 1}`,
                point_type: idx === 0 ? "ENTRY" : idx === coords.length - 1 ? "EXIT" : "WAYPOINT",
                latitude_dec: c[1],
                longitude_dec: c[0],
                min_alt_ft: props.floor_alt_ft_msl || 200,
                max_alt_ft: props.ceiling_alt_ft_msl || 10000,
                width_left_nm: (props.route_width_nm || 10.0) / 2.0,
                width_right_nm: (props.route_width_nm || 10.0) / 2.0,
              }));
              this.enrichAndSaveRoute({
                route_id: props.route_id,
                route_type: props.route_type,
                route_name: props.route_name,
                originating_agency: props.originating_agency,
                artcc_facility: props.artcc_facility,
                floor_alt_ft: props.floor_alt_ft_msl || 200,
                ceiling_alt_ft: props.ceiling_alt_ft_msl || 10000,
                route_width_nm: props.route_width_nm || 10.0,
                status: props.status || "ACTIVE",
                segments,
              });
            }
            loaded = true;
            console.log(`Loaded ${this.routes.size} routes from ${path.basename(p)} GeoJSON`);
            break;
          }
        }
      }

      if (!loaded || this.routes.size === 0) {
        console.warn("No initial routes loaded. Seeding default demo route.");
        this.enrichAndSaveRoute({
          route_id: "IR-102",
          route_type: "IR",
          route_name: "IR-102 Mojave Desert Low Altitude Corridor",
          originating_agency: "USAF / Edwards AFB",
          artcc_facility: "ZLA - Los Angeles ARTCC",
          floor_alt_ft: 100,
          ceiling_alt_ft: 15000,
          route_width_nm: 10.0,
          status: "ACTIVE",
          segments: [
            {
              sequence_num: 1,
              point_name: "PT_A_ENTRY",
              point_type: "ENTRY",
              latitude_dec: 34.9056,
              longitude_dec: -117.8821,
              min_alt_ft: 100,
              max_alt_ft: 15000,
              width_left_nm: 5.0,
              width_right_nm: 5.0,
            },
            {
              sequence_num: 2,
              point_name: "PT_B_TURN",
              point_type: "TURN_POINT",
              latitude_dec: 35.1211,
              longitude_dec: -117.3811,
              min_alt_ft: 200,
              max_alt_ft: 15000,
              width_left_nm: 5.0,
              width_right_nm: 5.0,
            },
            {
              sequence_num: 3,
              point_name: "PT_C_EXIT",
              point_type: "EXIT",
              latitude_dec: 35.4512,
              longitude_dec: -116.8912,
              min_alt_ft: 500,
              max_alt_ft: 15000,
              width_left_nm: 5.0,
              width_right_nm: 5.0,
            },
          ],
        });
      }
    } catch (err) {
      console.error("Failed to load initial data:", err);
    }
  }

  private enrichAndSaveRoute(rawRoute: any): MTRRoute {
    const routeId = String(rawRoute.route_id || "MTR").trim().toUpperCase();
    const rawSegments = rawRoute.segments || [];

    const enrichedSegments: MTRSegment[] = [];
    let cumulativeNm = 0.0;

    for (let i = 0; i < rawSegments.length; i++) {
      const seg: MTRSegment = { ...rawSegments[i] };
      const lat1 = seg.latitude_dec;
      const lon1 = seg.longitude_dec;

      let lat2 = seg.next_lat_dec;
      let lon2 = seg.next_lon_dec;
      if ((lat2 == null || lon2 == null) && i + 1 < rawSegments.length) {
        lat2 = rawSegments[i + 1].latitude_dec;
        lon2 = rawSegments[i + 1].longitude_dec;
        seg.next_lat_dec = lat2;
        seg.next_lon_dec = lon2;
        if (!seg.next_point_name) {
          seg.next_point_name = rawSegments[i + 1].point_name;
        }
      }

      let dist = seg.segment_distance_nm;
      if (lat2 != null && lon2 != null) {
        const bearing = calculateBearing(lat1, lon1, lat2, lon2);
        seg.bearing_deg = bearing;
        if (dist == null) {
          dist = calculateDistanceNm(lat1, lon1, lat2, lon2);
          seg.segment_distance_nm = dist;
        }
      }

      const distVal = dist != null ? dist : 0.0;
      seg.cumulative_distance_nm = Math.round(cumulativeNm * 10) / 10;
      cumulativeNm += distVal;

      const [arincLat, arincLon] = toArincDms(lat1, lon1);
      const [humanLat, humanLon] = toHumanDms(lat1, lon1);
      seg.arinc_lat = arincLat;
      seg.arinc_lon = arincLon;
      seg.human_lat = humanLat;
      seg.human_lon = humanLon;

      enrichedSegments.push(seg);
    }

    const corridorPoly = generateCorridorPolygon(enrichedSegments);

    const fullRoute: MTRRoute = {
      ...rawRoute,
      route_id: routeId,
      route_type: (rawRoute.route_type || "IR").toUpperCase(),
      route_name: rawRoute.route_name || `${routeId} Training Route`,
      originating_agency: rawRoute.originating_agency || "USAF / Command",
      artcc_facility: rawRoute.artcc_facility || "ARTCC",
      operating_hours: rawRoute.operating_hours || "CONTINUOUS",
      floor_alt_ft: Number(rawRoute.floor_alt_ft ?? 100),
      ceiling_alt_ft: Number(rawRoute.ceiling_alt_ft ?? 15000),
      route_width_nm: Number(rawRoute.route_width_nm ?? 10.0),
      status: rawRoute.status || "ACTIVE",
      segments: enrichedSegments,
      waypoint_count: enrichedSegments.length,
      total_distance_nm: Math.round(cumulativeNm * 10) / 10,
      corridor_polygon: corridorPoly,
    };

    this.routes.set(routeId, fullRoute);
    return fullRoute;
  }

  private validateRoute(routeData: any): void {
    if (!routeData || typeof routeData !== "object" || Array.isArray(routeData)) {
      throw new Error("Route payload must be a JSON object");
    }

    const routeId = String(routeData.route_id || "").trim().toUpperCase();
    if (!/^[A-Z0-9][A-Z0-9_-]{0,31}$/.test(routeId)) {
      throw new Error("route_id must be 1-32 characters using only letters, numbers, '_' or '-'");
    }

    const routeType = String(routeData.route_type || "").toUpperCase();
    if (!new Set(["IR", "VR", "SR"]).has(routeType)) {
      throw new Error("route_type must be IR, VR, or SR");
    }

    const requiredText = ["route_name", "originating_agency", "artcc_facility"];
    for (const field of requiredText) {
      if (typeof routeData[field] !== "string" || routeData[field].trim().length === 0 || routeData[field].length > 200) {
        throw new Error(`${field} must be a non-empty string of at most 200 characters`);
      }
    }

    const finiteInRange = (value: unknown, min: number, max: number) =>
      typeof value === "number" && Number.isFinite(value) && value >= min && value <= max;
    if (!finiteInRange(routeData.floor_alt_ft, 0, 99999) || !finiteInRange(routeData.ceiling_alt_ft, 0, 99999)) {
      throw new Error("Route altitudes must be finite numbers between 0 and 99,999 feet");
    }
    if (routeData.floor_alt_ft > routeData.ceiling_alt_ft) {
      throw new Error("floor_alt_ft cannot exceed ceiling_alt_ft");
    }
    if (!finiteInRange(routeData.route_width_nm, 0.1, 99.9)) {
      throw new Error("route_width_nm must be between 0.1 and 99.9 NM");
    }
    if (!Array.isArray(routeData.segments) || routeData.segments.length < 2 || routeData.segments.length > 1000) {
      throw new Error("segments must contain between 2 and 1,000 waypoints");
    }

    const sequences = new Set<number>();
    routeData.segments.forEach((segment: any, index: number) => {
      if (!segment || typeof segment !== "object" || Array.isArray(segment)) {
        throw new Error(`segments[${index}] must be an object`);
      }
      if (!Number.isInteger(segment.sequence_num) || segment.sequence_num < 1 || sequences.has(segment.sequence_num)) {
        throw new Error(`segments[${index}].sequence_num must be a unique positive integer`);
      }
      sequences.add(segment.sequence_num);
      if (typeof segment.point_name !== "string" || !/^[A-Za-z0-9_.-]{1,40}$/.test(segment.point_name)) {
        throw new Error(`segments[${index}].point_name contains unsupported characters`);
      }
      if (!new Set(["ENTRY", "WAYPOINT", "TURN_POINT", "EXIT"]).has(String(segment.point_type).toUpperCase())) {
        throw new Error(`segments[${index}].point_type is invalid`);
      }
      if (!finiteInRange(segment.latitude_dec, -90, 90) || !finiteInRange(segment.longitude_dec, -180, 180)) {
        throw new Error(`segments[${index}] coordinates are outside WGS84 bounds`);
      }
      for (const widthField of ["width_left_nm", "width_right_nm"] as const) {
        if (segment[widthField] != null && !finiteInRange(segment[widthField], 0, 99.9)) {
          throw new Error(`segments[${index}].${widthField} must be between 0 and 99.9 NM`);
        }
      }
      for (const altitudeField of ["min_alt_ft", "max_alt_ft"] as const) {
        if (segment[altitudeField] != null && !finiteInRange(segment[altitudeField], 0, 99999)) {
          throw new Error(`segments[${index}].${altitudeField} must be between 0 and 99,999 feet`);
        }
      }
      if (segment.min_alt_ft != null && segment.max_alt_ft != null && segment.min_alt_ft > segment.max_alt_ft) {
        throw new Error(`segments[${index}] minimum altitude cannot exceed maximum altitude`);
      }
    });
  }

  public listRoutes(routeType?: string): MTRRouteSummary[] {
    const list: MTRRouteSummary[] = [];
    const filterType = routeType ? routeType.trim().toUpperCase() : null;

    for (const r of this.routes.values()) {
      if (filterType && r.route_type !== filterType) continue;
      list.push({
        route_id: r.route_id,
        route_type: r.route_type,
        route_name: r.route_name,
        originating_agency: r.originating_agency,
        artcc_facility: r.artcc_facility,
        floor_alt_ft: r.floor_alt_ft,
        ceiling_alt_ft: r.ceiling_alt_ft,
        route_width_nm: r.route_width_nm,
        status: r.status || "ACTIVE",
        waypoint_count: r.segments.length,
      });
    }

    // Sort alphabetically by route_id
    list.sort((a, b) => a.route_id.localeCompare(b.route_id, undefined, { numeric: true }));
    return list;
  }

  public getRouteDetails(routeId: string): MTRRoute | null {
    return this.routes.get(routeId.trim().toUpperCase()) || null;
  }

  public getRouteGeojson(routeId: string): any | null {
    const route = this.getRouteDetails(routeId);
    if (!route) return null;

    const features: any[] = [];
    const segments = route.segments || [];

    // 1. Centerline LineString
    const coords: number[][] = [];
    for (const s of segments) {
      coords.push([s.longitude_dec, s.latitude_dec]);
    }
    if (segments.length > 0 && segments[segments.length - 1].next_lon_dec != null) {
      coords.push([
        segments[segments.length - 1].next_lon_dec!,
        segments[segments.length - 1].next_lat_dec!,
      ]);
    }

    features.push({
      type: "Feature",
      id: `${route.route_id}-centerline`,
      properties: {
        feature_type: "CENTERLINE",
        route_id: route.route_id,
        route_type: route.route_type,
        route_name: route.route_name,
        agency: route.originating_agency,
        floor_alt_ft: route.floor_alt_ft,
        ceiling_alt_ft: route.ceiling_alt_ft,
        width_nm: route.route_width_nm,
        total_distance_nm: route.total_distance_nm,
      },
      geometry: {
        type: "LineString",
        coordinates: coords,
      },
    });

    // 2. Corridor Ribbon Polygon
    const corridorCoords = route.corridor_polygon || [];
    if (corridorCoords.length > 0) {
      features.push({
        type: "Feature",
        id: `${route.route_id}-corridor`,
        properties: {
          feature_type: "CORRIDOR_RIBBON",
          route_id: route.route_id,
          width_nm: route.route_width_nm,
          floor_alt_ft: route.floor_alt_ft,
          ceiling_alt_ft: route.ceiling_alt_ft,
        },
        geometry: {
          type: "Polygon",
          coordinates: [corridorCoords],
        },
      });
    }

    // 3. Waypoint Points
    for (const s of segments) {
      features.push({
        type: "Feature",
        id: `${route.route_id}-${s.point_name}`,
        properties: {
          feature_type: "WAYPOINT",
          route_id: route.route_id,
          point_name: s.point_name,
          point_type: s.point_type,
          sequence_num: s.sequence_num,
          min_alt_ft: s.min_alt_ft,
          max_alt_ft: s.max_alt_ft,
          human_coords: `${s.human_lat}, ${s.human_lon}`,
          arinc_coords: `${s.arinc_lat}, ${s.arinc_lon}`,
        },
        geometry: {
          type: "Point",
          coordinates: [s.longitude_dec, s.latitude_dec],
        },
      });
    }

    return {
      type: "FeatureCollection",
      name: `MTR_${route.route_id}`,
      route_id: route.route_id,
      features,
    };
  }

  public getAllRoutesGeojson(): any {
    const allRoutes = this.listRoutes();
    const unifiedFeatures: any[] = [];
    for (const rSummary of allRoutes) {
      const rGeo = this.getRouteGeojson(rSummary.route_id);
      if (rGeo && Array.isArray(rGeo.features)) {
        unifiedFeatures.push(...rGeo.features);
      }
    }
    return {
      type: "FeatureCollection",
      name: "eNASR_All_Military_Training_Routes",
      features: unifiedFeatures,
    };
  }

  public getArinc424Xml(routeId?: string): string {
    if (routeId) {
      const route = this.getRouteDetails(routeId);
      if (!route) return `<!-- Route '${routeId}' not found -->`;
      return generateSingleRouteXml(route);
    } else {
      const routes = Array.from(this.routes.values());
      return generateArinc424Xml(routes);
    }
  }

  public getArinc424Fixed(routeId?: string): string {
    if (routeId) {
      const route = this.getRouteDetails(routeId);
      if (!route) return `// Route '${routeId}' not found\n`;
      return generateMtrFixedRecords(route).join("\n");
    } else {
      const routes = Array.from(this.routes.values());
      return generateAllFixedRecords(routes);
    }
  }

  public getPlainEnglish(routeId: string): string {
    const route = this.getRouteDetails(routeId);
    if (!route) return `Route '${routeId}' not found.`;
    return decodeToPlainEnglish(route);
  }

  public createOrUpdateRoute(routeData: any): MTRRoute {
    this.validateRoute(routeData);
    return this.enrichAndSaveRoute(routeData);
  }
}
